lncfg/address.go

```diff
@@ -0,0 +1,175 @@
+package lncfg
+
+import (
+	"fmt"
+	"strings"
+	"net"
+	"crypto/tls"
+	"time"
+)
+
+// NormalizeAddresses returns a new slice with all the passed addresses
+// normalized with the given default port and all duplicates removed.
+func NormalizeAddresses(addrs []string, defaultPort string) ([]string, error) {
+	result := make([]string, 0, len(addrs))
+	seen := map[string]struct{}{}
+	for _, strAddr := range addrs {
+		addr, err := parseAddressString(strAddr, defaultPort)
+		if err != nil {
+			return nil, err
+		}
+
+		if _, ok := seen[addr.String()]; !ok {
+			result = append(result, fmt.Sprintf("%s://%s",
+				addr.Network(), addr.String()))
+			seen[addr.String()] = struct{}{}
+		}
+	}
+	return result, nil
+}
+
+// EnforceSafeAuthentication enforces "safe" authentication taking into account
+// the interfaces that the RPC servers are listening on, and if macaroons are
+// activated or not. To protect users from using dangerous config combinations,
+// we'll prevent disabling authentication if the sever is listening on a public
+// interface.
+func EnforceSafeAuthentication(addrs []string, macaroonsActive bool) error {
+
+	// We'll now examine all addresses that this RPC server is listening
+	// on. If it's a localhost address, we'll skip it, otherwise, we'll
+	// return an error if macaroons are inactive.
+	for _, addr := range addrs {
+		if IsLoopback(addr) || IsUnix(addr) {
+			continue
+		}
+
+		if !macaroonsActive {
+			return fmt.Errorf("Detected RPC server listening on "+
+				"publicly reachable interface %v with "+
+				"authentication disabled! Refusing to start "+
+				"with --no-macaroons specified.", addr)
+		}
+	}
+
+	return nil
+}
+
+// SplitNormalizedAddress splits a normalized address into the network and the
+// address part. A normalized address looks like this: network://address[:port]
+func SplitNormalizedAddress(strAddress string) (string, string, error) {
+	parts := strings.Split(strAddress, "://")
+	if len(parts) != 2 {
+		return "", "", fmt.Errorf("Invalid normalized address: %s",
+			strAddress)
+	}
+	return parts[0], parts[1], nil
+}
+
+// ListenOnNormalizedAddress creates a listener that listens on the given
+// address. The address must be in the normalized format that is returned
+// by normalizeAddresses, which is network://address[:port].
+// Valid examples:
+//	tcp://localhost:9735
+//	tcp6://[::1]:9735
+//	unix:///tmp/listen.sock
+func ListenOnNormalizedAddress(strAddress string) (net.Listener, error) {
```

Comment:
Instead of first needing to normalize the address, then splitting it here (meaning we go from `string` -> `net.Addr` -> `string`) it would make sense to let the "normalization" process just return a list of `net.Addr`s, which could be listened on.